src/jit/importer.cpp

```diff
@@ -4122,7 +4122,8 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
                 {
                     case CorInfoType::CORINFO_TYPE_SHORT:
                     case CorInfoType::CORINFO_TYPE_USHORT:
-                        retNode = gtNewOperNode(GT_BSWAP16, callType, impPopStack().val);
+                        retNode = gtNewCastNode(TYP_INT, gtNewOperNode(GT_BSWAP16, TYP_INT, impPopStack().val), false,
```

Comment:
I don't think we currently have this for any of the HWIntrinsics? Are we doing something special there or are we missing these in that case as well?

Reply:
Yeah, that's something we need to check. I'm pretty sure I've seen this issue in the past but I don't remember where.

Reply:
What I remember is probably `pextrb`/`pextrw` but those were fixed in a different manner, by removing the signed versions.

Reply:
If `PEXTRW` is ok for unsigned without an extra cast, wouldn't `BSWAP16` also be?  Since we get a sign/zero extended load before the `SBWAP16` (because it's defined as operating on `TYP_INT`?), I can only see there being a problem when the intrinsic changes the sign.  Otherwise we get 

```asm
movzx rax reg.16/mem
ror ax, 8
movxz rax ax
```

where the second zero-extend is most definitely redundant.

Reply:
Nope, `pextrb/w` automatically zero extend the value to 32 bit, `ror` does not do that.

> I can only see there being a problem when the intrinsic changes the sign. 

Exactly. It's the cast before `BSWAP16` that's redundant, the one after `BSWAP16` is redundant if only the lower 16 bits are consumed. For example, when the result of `BSWAP16` is stored in a 16 bit memory location. There's code in the JIT that tries to do this but it does not handle all cases.

Reply:
> I don't think we currently have this for any of the HWIntrinsics?

I have added the cast for `GetElement` on small int https://github.com/dotnet/coreclr/pull/22353